### ISSUE TITLE
no-unnecessary-type-assertion: false positives where tsc requires the assertion

### ISSUE THREAD

**Issue by assapir** [oxc-project/tsgolint#669]
## Description

`no-unnecessary-type-assertion` flags several type assertions as unnecessary, but removing them causes `tsc` to fail with type errors. These are false positives where tsgolint's type analysis disagrees with `tsc 5.9.3`.

All cases below are **verified with `tsc 5.9.3 --strict`** — with the assertion: passes. Without: fails.

## Environment

- oxlint: 1.46.0
- oxlint-tsgolint: 0.12.0
- TypeScript: 5.9.3
- OS: macOS (darwin-arm64), CI on Linux (linux-arm64)

### Case 1: `PropertyKey` to `string`

```typescript
function filterProps(props: PropertyKey[]): string[] {
  return props.filter((prop) =>
    !["foo", "bar"].includes(prop as string)
    //                       ^^^^^^^^^^^
    // tsgolint: "This assertion is unnecessary"
    // Without assertion → TS2345: Argument of type 'PropertyKey' is not
    //   assignable to parameter of type 'string'.
    //   Type 'number' is not assignable to type 'string'.
  ) as string[];
}
```

`PropertyKey` is `string | number | symbol`. Casting to `string` narrows the type for `Array<string>.includes()`.

### Case 2: `unknown` to `Record<string, unknown>`

```typescript
async function mergeWithDefaults(loadModule: () => Promise<unknown>) {
  const mod = (await loadModule()) as Record<string, unknown>;
  //          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  // tsgolint: "This assertion is unnecessary"
  // Without assertion → TS2698: Spread types may only be created from object types.
  return { ...mod, extra: true };
}
```

`unknown` cannot be spread without assertion.

### Case 3: Generic union not narrowed by `typeof`

```typescript
type Wrapper<T> = { value: number; meta: T };

function unwrap<T>(input: number | Wrapper<T>): number {
  return typeof input === "string" ? parseFloat(input) : (input as number);
  //                                                      ^^^^^^^^^^^^^^^
  // tsgolint: "This assertion is unnecessary"
  // Without assertion → TS2322: Type 'number | Wrapper<T>' is not assignable
  //   to type 'number'. Type 'Wrapper<T>' is not assignable to type 'number'.
}
```

`typeof input === "string"` doesn't narrow away `Wrapper<T>` in the else branch because `T` is generic, so the type remains `number | Wrapper<T>`.

## Verification

```bash
# All pass with assertions
tsc --noEmit --strict case1.ts case2.ts case3.ts  # exit 0

# All fail without assertions
tsc --noEmit --strict case1-no-assert.ts  # TS2345
tsc --noEmit --strict case2-no-assert.ts  # TS2698
tsc --noEmit --strict case3-no-assert.ts  # TS2322
```

## Real-world impact

These patterns caused ~61 false positives across our monorepo:

| Case | Real-world pattern | Count |
|---|---|---|
| 1 | MUI `shouldForwardProp(prop)` where `prop: PropertyKey` | 11 files |
| 2 | Vitest `vi.mock(_, async (importOriginal) => { const actual = (await importOriginal()) as Record<string, unknown>; return { ...actual }; })` | 36 files |
| 3 | Generic components with union type props | 4 files |

## Expected behavior

All cases should NOT be flagged — removing the assertion causes `tsc 5.9.3` to produce type errors.

## Actual behavior

tsgolint flags all as "This assertion is unnecessary since it does not change the type of the expression."

**Comment by camc314** [comment]
Thanks for using oxlint/tsgolint. 

Please try 0.12.1 and let me know if this is fixed - thanks

**Comment by assapir** [comment]
Thanks for the fix in 0.12.1 — Cases 1 (`PropertyKey` → `string`) and 2 (`unknown` → `Record<string, unknown>`) from the original report are now resolved!

However, there's still a false positive with **generic conditional types with unions**. This is similar to Case 3 from the original report but with a conditional type.

## Environment

- oxlint: 1.47.0
- oxlint-tsgolint: 0.12.1
- TypeScript: 5.9.3

## Minimal reproduction (no dependencies)

```typescript
type NumberValueType = number | string;
type NumberValuePairType = [NumberValueType, NumberValueType];

type NumberCellValueType<T extends NumberValuePairType | NumberValueType> =
  T extends NumberValuePairType ? NumberValuePairType : NumberValueType;

function processValue<T extends NumberValuePairType | NumberValueType>(
  value: NumberCellValueType<T>
): number {
  if (Array.isArray(value)) {
    return 0;
  }

  const numberValue = typeof value === "string" ? parseFloat(value) : (value as number);
  //                                                                   ^^^^^^^^^^^^^^^^
  // tsgolint: "This assertion is unnecessary since it does not change the type of the expression."
  const negative = numberValue < 0;
  return negative ? -1 : 1;
}
```

## Verification

```bash
# WITH assertion — tsc passes
tsc --noEmit --strict repro.ts  # exit 0

# WITHOUT assertion — tsc fails
# TS2365: Operator '<' cannot be applied to types 'number | NumberCellValueType<T>' and 'number'.
```

The generic conditional type `NumberCellValueType<T>` isn't fully resolved when `T` is generic. After `Array.isArray` and `typeof === "string"` guards, tsc sees the remaining type as `number | NumberCellValueType<T>` (not narrowed to just `number`), so the `as number` assertion is required.

**Comment by assapir** [comment]
Follow-up: Case 2 from the original report (`unknown` → `Record<string, unknown>`) is also **not fixed** in 0.12.1.

## Additional minimal reproduction — spread of `unknown`

```typescript
async function mergeWithDefaults(loadModule: () => Promise<unknown>) {
  const mod = (await loadModule()) as Record<string, unknown>;
  //          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  // tsgolint 0.12.1: "This assertion is unnecessary"
  // Without assertion: TS2698: Spread types may only be created from object types.
  return { ...mod, extra: true };
}
```

```bash
# WITH assertion — passes
tsc --noEmit --strict spread-repro.ts  # exit 0

# WITHOUT assertion — fails
# TS2698: Spread types may only be created from object types.
```

This pattern is very common in Vitest mocks:
```typescript
vi.mock("react-router-dom", async (importOriginal) => {
  const actual = (await importOriginal()) as Record<string, unknown>;
  return { ...actual, useLocation: vi.fn() };
});
```

So to summarize, from the original 3 cases in this issue:
- **Case 1** (`PropertyKey` → `string`): Fixed in 0.12.1 ✅
- **Case 2** (`unknown` → `Record<string, unknown>`): Still false positive ❌
- **Case 3** (generic union not narrowed by `typeof`): Still false positive ❌ (see previous comment)

**Comment by camc314** [comment]
Hopefully i've got them all this time 😅 

Will ship 0.12.2 soon

**Comment by assapir** [comment]
Confirmed to work on our repo as expected
